dist/functional-test.sh

```diff
@@ -0,0 +1,155 @@
+#!/bin/bash
+
+
+ETCD_IMG="quay.io/coreos/etcd:v3.0.3"
+FLANNEL_NET="10.10.0.0/16"
+
+usage() {
+	echo "$0 FLANNEL-DOCKER-IMAGE"
+	echo
+	echo "Run end-to-end tests by bringing up two flannel instances"
+	echo "and having them ping each other"
+	echo
+	echo "NOTE: this script depends on Docker 1.9.0 or higher"
+	exit 1
+}
+
+version_check() {
+	required=$1
+	actual=$2
+	err_prefix=$3
+
+	req_maj=$(echo $required | cut -d . -f 1)
+	req_min=$(echo $required | cut -d . -f 2)
+	act_maj=$(echo $actual | cut -d . -f 1)
+	act_min=$(echo $actual | cut -d . -f 2)
+
+	if [ $act_maj -lt $req_maj ] || ( [ $act_maj -eq $req_maj ] && [ $act_min -lt $req_min ] ); then
+		echo "$err_prefix: required=$required, found=$actual"
+		exit 1
+	fi
+}
+
+docker_version_check() {
+	ver=$(docker version -f '{{.Server.Version}}')
+	version_check "1.9" $ver
+}
+
+run_test() {
+	backend=$1
+
+	while true; do
+		# etcd might take a bit to come up
+		docker run -it --name=flannel-test-etcdctl --entrypoint=/usr/local/bin/etcdctl $ETCD_IMG \
+			--endpoints=$etcd_endpt "set" /coreos.com/network/config "{ \"Network\": \"$FLANNEL_NET\", \"Backend\": { \"Type\": \"${backend}\" } }"
+
+		ok=$?
+
+		docker rm "flannel-test-etcdctl" >/dev/null
+
+		if [ $ok -eq 0 ]; then
+			break
+		fi
+
+		sleep 5
+	done
+
+	echo flannel config written
+
+	flannel1=$(docker run -d --privileged --entrypoint=/opt/bin/flanneld $flannel_img --etcd-endpoints=$etcd_endpt)
+	if [ $? -ne 0 ]; then
+		exit 1
+	fi
+
+	flannel2=$(docker run -d --privileged --entrypoint=/opt/bin/flanneld $flannel_img --etcd-endpoints=$etcd_endpt)
+	if [ $? -ne 0 ]; then
+		exit 1
+	fi
+
+	echo flannels running
+	sleep 5
+
+	# add a dummy interface with $FLANNEL_SUBNET so we have a known working IP to ping
+	ping_dest=$(docker "exec" --privileged $flannel1 /bin/sh -c '\
+		source /run/flannel/subnet.env && 
+		ip link add name dummy0 type dummy && \
+		ip addr add $FLANNEL_SUBNET dev dummy0 && \
+	       	ip link set dummy0 up && \
+		echo $FLANNEL_SUBNET | cut -f 1 -d "/" ')
+
+	docker "exec" -it --privileged $flannel2 /bin/ping -c 5 $ping_dest
```

Comment:
I was thinking that `exec` is a bash keyword but it's just a builtin. I'll drop it.